File: DSA/searching/jump_search.cpp
#include <iostream>
#include <cmath>

using namespace std;

double min(double x1, double x2){
    if (x1 < x2){
        return x1;
    }else{
        return x2;
    }
}

int jump_search(int arr[], int x, int n)
{
    int step = sqrt(n);

    int prev = 0;
    while (arr[min(step, n)-1] < x)
    {
        prev = step;
        step += sqrt(n);
        if (prev >= n)
            return -1;
    }
 
    while (arr[prev] < x){
        prev++;
        if (prev == min(step, n))
            return -1;
    }

    if (arr[prev] == x)
        return prev;
    return -1;
}

int main(){
    int array[10] = {1,2,3,4,5,6,7,8,9,10};

    cout << jump_search(array, 10, 4) << endl;
    cout << jump_search(array, 10, 10) << endl;
    cout << jump_search(array, 10, 0) << endl;
    cout << jump_search(array, 10, 55) << endl;
    cout << jump_search(array, 10, 2) << endl;
}

File: DSA/searching/jump_search.py
import math

def jump_search( arr , x , n ):
    step = math.sqrt(n) 
    prev = 0
    while arr[int(min(step, n)-1)] < x:
        prev = step
        step += math.sqrt(n)
        if prev >= n:
            return -1
     
    # Doing a linear search for x in
    # block beginning with prev.
    while arr[int(prev)] < x:
        prev += 1
         
        # If we reached next block or end
        # of array, element is not present.
        if prev == min(step, n):
            return -1
     
    # If element is found
    if arr[int(prev)] == x:
        return prev
     
    return -1

if __name__ == "__main__":
    lista = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    element = 42
    print(jump_search(lista, element, len(lista)))